File: Source/AlphaTab/Util/Lazy.cs
/*
 * This file is part of alphaTab.
 * Copyright © 2017, Daniel Kuschny and Contributors, All rights reserved.
 * 
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3.0 of the License, or at your option any later version.
 * 
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library.
 */
using System;

namespace AlphaTab.Util
{
    public class Lazy<T>
    {
        private readonly Func<T> _factory;
        private bool _created;
        private T _value;

        public Lazy(Func<T> factory)
        {
            _factory = factory;
        }

        public T Value
        {
            get
            {
                if (!_created)
                {
                    _value = _factory();
                    _created = true;
                }
                return _value;
            }
        }
    }
}

File: packages/kotlin/src/android/src/main/java/alphaTab/util/Lazy.kt
package alphaTab.util

internal object UninitializedValue

internal class Lazy<T : Any?>(factory: () -> T) {
    private val _factory: () -> T = factory
    private var _value:Any? = UninitializedValue

    public val value:T
        get() {
            if(_value == UninitializedValue) {
                _value = _factory()
            }
            @Suppress("UNCHECKED_CAST")
            return _value as T
        }
}
